ledger-api/grpc-definitions/com/daml/ledger/api/v1/error.proto

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+syntax = "proto3";
+
+package com.daml.ledger.api.v1;
+
+// Rules on error reporting over the Ledger API
+// ********************************************
+//
+// 1. We use gRPC status codes to report synchronous success and failures
+//    coming from the underlying gRPC protocol or the API-level request processing.
+//    Asynchronous API-level errors are reported over a completion stream like the command completion stream,
+//    which themselves include a ``google.rpc.Status`` message to indicate the status.
+//    The status code ``OK`` denotes success.
+//
+// 2. Error reporting includes relevant error details in the ``details`` field of the status message,
+//    following the gRPC rich error model as described at https://cloud.google.com/apis/design/errors#error_model.
+//    Error details use the `google.rpc.ErrorInfo` message defined at
+//    https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto#L112.
+//    Its ``reason`` field includes an alphanumeric error code.
+//    Further details on the error and information for subsequent API calls may be included in the ``metadata`` field.
+//    The Ledger API RPC documentation may document further metadata of the error.
+//
+// 3. RPC requests can specify that the API-level error details of the RPC response shall be JSON-encoded in the error message
+//    instead of using the ``details`` field. To that end, the RPC request shall specify the ``X-Daml-Error-Details-In-Message``
+//    HTTP/2 header in the request. If an application cannot read the error details from the trailers,
+//    e.g. because some middleware drops them, it can obtain the details by parsing the message.
+//    This flag does not affect asynchronous errors reported as part of completions.
+//
+//    TODO(v2) Evaluate whether we can implement this as a gRPC interceptor that listens for the header
+//     and converts the rich error details into JSON. Otherwise this might not work for
+//     errors coming from the underlying gRPC protocol.
+//     If this doesn't work, we can alternatively add a flag to all RPC requests.
+//     For transparency, consider encoding the whole ``google.rpc.Status message`` in the message.
+//
+//    TODO(v2) We shoule use the same JSON encoding here as what we get from our JSON API server.
```

Comment:
Not quite sure what you mean here. We don’t have structured errors on the JSON API atm. Errors are an http status code + a list of string error messages.

Reply:
With the new proposal, applications will need to obtain some information from the errors that they can use for retries or further submissions (mostly offsets). So what does the JSON API server do if it gets a gRPC error with rich error details? How do you pass this information on to the client?

Reply:
Ah I misread this as suggesting to match the existing error format. Matching how we represent the rich errors seems very sensible :+1: I’m hoping to find the time today to write up a proposal for this.